File: scala/src/main/scala/Day04.scala
import scala.collection.parallel.CollectionConverters._
import scala.io.Source

object Day04 extends App {
  val filename = "../input04"

  val byr = (value:String) => {
    1920 <= value.toInt && value.toInt <= 2002
  }

  val iyr = (value:String) => {
    2010 <= value.toInt && value.toInt <= 2020
  }

  val eyr = (value:String) => {
    2020 <= value.toInt && value.toInt <= 2030
  }

  val hgt = (value:String) => {
    value match {
      case x if x.endsWith("in") => {
        val v = x.init.init.toInt
        59<=v && v <= 76
      }
      case x if x.endsWith("cm") =>{
        val v = x.init.init.toInt
        150 <=v && v <= 193
      }
      case _=> false
    }
  }

  val hcl = (value:String) => value.matches("#([0-9]|[a-f]){6}")

  val ecl = (value:String) => Set("amb","blu","brn","gry","grn","hzl","oth").contains(value)

  val pid = (value:String) => value.matches("[0-9]{9}")

  val cid = (value:String) => true

  val baseValidField = Map("byr" -> byr,"iyr" -> iyr,"eyr" -> eyr,"hgt" ->hgt,"hcl" -> hcl,"ecl" -> ecl,"pid" -> pid)
  val baseValidField2 = baseValidField + ("cid" -> cid)
  val validField = Seq(
    baseValidField,
    baseValidField2)


  val s = Source.fromFile(filename)
  val map = s.
    getLines()
    .mkString(System.lineSeparator)
    .split(System.lineSeparator*2)
    .toSeq
    .count {
      id =>
        val s = id.replace(System.lineSeparator, " ")
        .split(" ")
        .toSeq
        .map(_.split(":"))
        .toSet
        validField.map(_.keys.toSet).contains(s.map(_.head)) && s.forall {
            case Array(key, value) =>
              baseValidField2(key)(value)
          }

    }
  println(map)
  s.close()
}

File: rust/aoc/src/day04.rs
use std::collections::HashSet;
use std::fs;

pub fn day04() {
    fn parse_check(s: String) -> i32 {
        let valid_keys: HashSet<String> = [
            String::from("byr"),
            String::from("iyr"),
            String::from("eyr"),
            String::from("hgt"),
            String::from("hcl"),
            String::from("ecl"),
            String::from("pid"),
        ]
        .iter()
        .cloned()
        .collect::<HashSet<_>>();
        let mut string_iterator = s.chars().peekable();
        let mut valid_key_count = 7;
        let mut is_value_valid = true;
        while string_iterator.peek().is_some() {
            let key = string_iterator
                .by_ref()
                .take_while(|&x| x != ':')
                .collect::<String>();
            let value = string_iterator
                .by_ref()
                .take_while(|&x| x != ' ')
                .collect::<String>();
            if valid_keys.contains(&key) {
                valid_key_count -= 1;
                match &key[..] {
                    "byr" => {
                        is_value_valid &= 1920 <= value.as_str().parse::<i32>().unwrap()
                            && value.as_str().parse::<i32>().unwrap() <= 2002
                    }
                    "iyr" => {
                        is_value_valid &= 2010 <= value.as_str().parse::<i32>().unwrap()
                            && value.as_str().parse::<i32>().unwrap() <= 2020
                    }
                    "eyr" => {
                        is_value_valid &= 2020 <= value.as_str().parse::<i32>().unwrap()
                            && value.as_str().parse::<i32>().unwrap() <= 2030
                    }
                    "hgt" => {}
                    "hcl" => {}
                    "ecl" => match value.as_str() {
                        "amb" | "blu" | "brn" | "gry" | "grn" | "hzl" | "oth" => (),
                        _ => is_value_valid = false,
                    },
                    "pid" => {
                        is_value_valid &= value
                            .chars()
                            .by_ref()
                            .take_while(|&x| x.is_digit(10) == false)
                            //.collect()
                            .len()
                            == 0
                    }
                    _ => (),
                }
            }
        }
        if valid_key_count <= 0 && is_value_valid {
            1
        } else {
            0
        }
    }
    let contents =
        fs::read_to_string("..\\..\\input04").expect("Something went wrong reading the file");
    let map: i32 = contents
        .split("\n\n")
        .map(|x| parse_check(x.replace("\n", " ")))
        .collect::<Vec<_>>()
        .iter()
        .sum();

    println!("{:?}", map);
}
